File: win/engine/send_queue.go
package engine

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type queuedSend struct {
	ID        string    `json:"id"`
	SessionID string    `json:"sessionId"`
	PeerID    string    `json:"peerId"`
	PeerName  string    `json:"peerName"`
	Host      string    `json:"host"`
	Port      uint16    `json:"port"`
	Paths     []string  `json:"paths"`
	Text      string    `json:"text"`
	Title     string    `json:"title"`
	Total     uint64    `json:"total"`
	FileCount int       `json:"fileCount"`
	LastTry   time.Time `json:"-"`
}

func sendQueuePath() string {
	return filepath.Join(filepath.Dir(settingsPath()), "send-queue.json")
}

func loadSendQueue(deviceID string) []queuedSend {
	if deviceID == "selftest-node" {
		return nil
	}
	raw, err := os.ReadFile(sendQueuePath())
	if err != nil {
		return nil
	}
	var list []queuedSend
	if json.Unmarshal(raw, &list) != nil {
		return nil
	}
	return list
}

func (n *Node) persistSendQueue() {
	if n.settingsCopy().isSelfTest() {
		return
	}
	n.mu.Lock()
	list := append([]queuedSend(nil), n.sendQueue...)
	n.mu.Unlock()
	raw, err := json.MarshalIndent(list, "", "  ")
	if err != nil {
		return
	}
	_ = os.MkdirAll(filepath.Dir(sendQueuePath()), 0o755)
	_ = os.WriteFile(sendQueuePath(), raw, 0o644)
}

func (n *Node) restoreQueuedItems() {
	n.mu.Lock()
	jobs := append([]queuedSend(nil), n.sendQueue...)
	have := map[string]bool{}
	for _, it := range n.items {
		have[it.ID] = true
	}
	n.mu.Unlock()
	for _, job := range jobs {
		if have[job.ID] {
			continue
		}
		n.addItem(TransferItem{
			ID: job.ID, SessionID: job.SessionID, Direction: "send", PeerName: job.PeerName,
			PeerID: job.PeerID, Title: job.Title, Total: job.Total, State: "等待对方上线",
			Detail: "网络中断，会自动接着传（已传过的部分会跳过）", FileCount: job.FileCount, Time: time.Now(),
			LocalPath: firstPath(job.Paths),
		})
	}
}

func firstPath(paths []string) string {
	if len(paths) == 0 {
		return ""
	}
	return paths[0]
}

func (n *Node) canQueue(p Peer) bool {
	return p.ID != "" && p.ID != "phone-web"
}

func (n *Node) shouldQueue(p Peer) bool {
	return n.canQueue(p) && !peerOnline(p, time.Now())
}

func isUnreachable(err error) bool {
	if err == nil {
		return false
	}
	s := err.Error()
	if strings.Contains(s, "分块多次校验失败") || strings.Contains(s, "对方还是旧版") ||
		strings.Contains(s, "协议不匹配") || strings.Contains(s, "对方拒绝") ||
		strings.Contains(s, "已取消") {
		return false
	}
	keys := []string{
		"连不上对方", "连接超时", "连接已断开", "等待数据超时", "数据通道",
		"握手失败", "对方没有回应",
		"connection refused", "i/o timeout", "broken pipe", "connection reset",
		"network is unreachable", "no route to host", "host is down",
		"use of closed network connection", "forcibly closed", "wsasend", "wsarecv",
		"software caused connection abort", "connection timed out",
	}
	low := strings.ToLower(s)
	for _, k := range keys {
		if strings.Contains(low, strings.ToLower(k)) {
			return true
		}
	}
	return false
}

func (n *Node) enqueueLocked(job queuedSend) {
	for i, old := range n.sendQueue {
		if old.ID == job.ID {
			n.sendQueue[i] = job
			return
		}
	}
	n.sendQueue = append(n.sendQueue, job)
}

func (n *Node) enqueue(job queuedSend) {
	n.mu.Lock()
	n.enqueueLocked(job)
	n.mu.Unlock()
	n.persistSendQueue()
}

func (n *Node) dropQueued(id, session string) {
	n.mu.Lock()
	out := n.sendQueue[:0]
	for _, job := range n.sendQueue {
		if job.ID == id || job.SessionID == session {
			continue
		}
		out = append(out, job)
	}
	n.sendQueue = out
	n.mu.Unlock()
	n.persistSendQueue()
}

func (n *Node) takeReadyJob(peerID string) (queuedSend, bool) {
	n.mu.Lock()
	defer n.mu.Unlock()
	now := time.Now()
	for i, job := range n.sendQueue {
		if job.PeerID != peerID {
			continue
		}
		if !job.LastTry.IsZero() && now.Sub(job.LastTry) < 20*time.Second {
			continue
		}
		n.sendQueue = append(n.sendQueue[:i], n.sendQueue[i+1:]...)
		return job, true
	}
	return queuedSend{}, false
}

func (n *Node) queueFiles(peer Peer, files []LocalFile, itemID, sessionID string) {
	if itemID == "" {
		itemID = newUUID()
	}
	if sessionID == "" {
		sessionID = newUUID()
	}
	var total uint64
	paths := make([]string, len(files))
	for i, f := range files {
		total += f.Size
		paths[i] = f.Path
	}
	title := files[0].Rel
	if len(files) > 1 {
		title = fmtCount(len(files))
	}
	n.updateOrAddWaiting(TransferItem{
		ID: itemID, SessionID: sessionID, Direction: "send", PeerName: peer.Name, PeerID: peer.ID,
		Title: title, Total: total, State: "等待对方上线", Detail: "网络中断，会自动接着传（已传过的部分会跳过）",
		FileCount: len(files), Time: time.Now(), LocalPath: files[0].Path,
	})
	n.enqueue(queuedSend{
		ID: itemID, SessionID: sessionID, PeerID: peer.ID, PeerName: peer.Name,
		Host: peer.Host, Port: peer.Port, Paths: paths, Title: title,
		Total: total, FileCount: len(files),
	})
}

func (n *Node) queueText(peer Peer, text string, itemID, sessionID string, cooldown bool) {
	if itemID == "" {
		itemID = newUUID()
	}
	if sessionID == "" {
		sessionID = newUUID()
	}
	title := text
	if len([]rune(title)) > 36 {
		title = string([]rune(title)[:36]) + "…"
	}
	n.updateOrAddWaiting(TransferItem{
		ID: itemID, SessionID: sessionID, Direction: "send", PeerName: peer.Name, PeerID: peer.ID,
		Title: title, State: "等待对方上线", Detail: "网络中断，会自动接着传（已传过的部分会跳过）", Time: time.Now(),
	})
	job := queuedSend{
		ID: itemID, SessionID: sessionID, PeerID: peer.ID, PeerName: peer.Name,
		Host: peer.Host, Port: peer.Port, Text: text, Title: title,
	}
	if cooldown {
		job.LastTry = time.Now()
	}
	n.enqueue(job)
}

func (n *Node) updateOrAddWaiting(it TransferItem) {
	n.mu.Lock()
	for i := range n.items {
		if n.items[i].ID == it.ID {
			n.items[i].State = it.State
			n.items[i].Detail = it.Detail
			n.items[i].Speed = 0
			n.mu.Unlock()
			return
		}
	}
	n.items = append([]TransferItem{it}, n.items...)
	n.mu.Unlock()
}

func fmtCount(n int) string {
	return fmt.Sprintf("%d 个文件", n)
}

func (n *Node) flushOnlineQueued() {
	n.mu.Lock()
	ids := map[string]bool{}
	for _, job := range n.sendQueue {
		ids[job.PeerID] = true
	}
	n.mu.Unlock()
	now := time.Now()
	for id := range ids {
		p, ok := n.PeerByID(id)
		if !ok || !peerOnline(p, now) {
			continue
		}
		go n.flushSendQueue(p)
	}
}

func (n *Node) flushSendQueue(peer Peer) {
	if peer.ID == "" || peer.ID == "phone-web" {
		return
	}
	n.mu.Lock()
	if n.flushing[peer.ID] {
		n.mu.Unlock()
		return
	}
	n.flushing[peer.ID] = true
	n.mu.Unlock()
	defer func() {
		n.mu.Lock()
		delete(n.flushing, peer.ID)
		n.mu.Unlock()
	}()
	for {
		job, ok := n.takeReadyJob(peer.ID)
		if !ok {
			n.persistSendQueue()
			return
		}
		n.persistSendQueue()
		if n.isCancelled(job.SessionID) {
			continue
		}
		live, found := n.PeerByID(peer.ID)
		if !found {
			live = peer
		}
		if !peerOnline(live, time.Now()) {
			job.Host, job.Port = live.Host, live.Port
			n.enqueue(job)
			return
		}
		n.runQueued(live, job)
	}
}

func (n *Node) runQueued(peer Peer, job queuedSend) {
	if job.Text != "" {
		n.deliverText(peer, job.Text, job.ID, job.SessionID)
		return
	}
	files, err := Collect(job.Paths)
	if err != nil {
		n.updateItem(job.ID, func(it TransferItem) TransferItem {
			it.State = "失败"
			it.Detail = "文件已经不在了"
			it.Speed = 0
			return it
		})
		return
	}
	n.sendFiles(peer, files, &TransferItem{ID: job.ID, SessionID: job.SessionID})
}

func (n *Node) requeueUnreachable(peer Peer, files []LocalFile, itemID, sessionID string) {
	job := queuedSend{
		ID: itemID, SessionID: sessionID, PeerID: peer.ID, PeerName: peer.Name,
		Host: peer.Host, Port: peer.Port, Title: "", LastTry: time.Now(),
	}
	if len(files) > 0 {
		job.Paths = make([]string, len(files))
		var total uint64
		for i, f := range files {
			job.Paths[i] = f.Path
			total += f.Size
		}
		job.Total = total
		job.FileCount = len(files)
		job.Title = files[0].Rel
		if len(files) > 1 {
			job.Title = fmtCount(len(files))
		}
	}
	n.enqueue(job)
	n.updateItem(itemID, func(it TransferItem) TransferItem {
		it.State = "等待对方上线"
		it.Detail = "网络中断，会自动接着传（已传过的部分会跳过）"
		it.Speed = 0
		return it
	})
}

File: Sources/HuChuan/SendQueue.swift
import Foundation
import HuChuanCore

struct QueuedSend: Equatable {
    var itemId: UUID
    var sessionId: String
    var peerId: String
    var peerName: String
    var host: String
    var port: UInt16
    var paths: [String]
    var text: String
    var title: String
    var total: UInt64
    var fileCount: Int
    var lastAttempt: Date = .distantPast
}

private struct QueuedSendDisk: Codable {
    var itemId: String
    var sessionId: String
    var peerId: String
    var peerName: String
    var host: String
    var port: UInt16
    var paths: [String]
    var text: String
    var title: String
    var total: UInt64
    var fileCount: Int
}

enum SendReachability {
    static let resumeHint = "网络中断，会自动接着传（已传过的部分会跳过）"

    static func isUnreachable(_ error: Error) -> Bool {
        let s = ((error as? AppError)?.description ?? "") + " " + error.localizedDescription + " " + String(describing: error)
        if s.contains("分块多次校验失败") || s.contains("对方还是旧版") ||
            s.contains("协议不匹配") || s.contains("对方拒绝") ||
            s.contains("连接已取消") || s.contains("已取消") {
            return false
        }
        let keys = [
            "连接超时", "连不上", "连接已断开", "等待数据超时", "等待超时", "数据通道",
            "握手失败", "对方没有回应", "Connection refused", "Could not connect",
            "couldn't connect", "Network is unreachable", "No route to host",
            "Host is down", "Socket is not connected", "Broken pipe",
            "Connection reset", "Software caused connection abort",
            "POSIXErrorCode(61)", "POSIXErrorCode(60)", "POSIXErrorCode(54)",
            "POSIXErrorCode(32)", "POSIXErrorCode(57)", "POSIXErrorCode(51)",
            "POSIXErrorCode(65)", "POSIXErrorCode(50)",
        ]
        return keys.contains { s.localizedCaseInsensitiveContains($0) }
    }
}

enum SendQueueFile {
    static func url() -> URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSHomeDirectory())
        let folder = dir.appendingPathComponent("互传", isDirectory: true)
        try? FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        return folder.appendingPathComponent("send-queue.json")
    }

    static func load() -> [QueuedSend] {
        guard let data = try? Data(contentsOf: url()),
              let list = try? JSONDecoder().decode([QueuedSendDisk].self, from: data) else {
            return []
        }
        return list.compactMap { row in
            guard let id = UUID(uuidString: row.itemId) else { return nil }
            return QueuedSend(
                itemId: id, sessionId: row.sessionId, peerId: row.peerId, peerName: row.peerName,
                host: row.host, port: row.port, paths: row.paths, text: row.text,
                title: row.title, total: row.total, fileCount: row.fileCount
            )
        }
    }

    static func save(_ items: [QueuedSend]) {
        let rows = items.map {
            QueuedSendDisk(
                itemId: $0.itemId.uuidString, sessionId: $0.sessionId, peerId: $0.peerId,
                peerName: $0.peerName, host: $0.host, port: $0.port, paths: $0.paths,
                text: $0.text, title: $0.title, total: $0.total, fileCount: $0.fileCount
            )
        }
        guard let data = try? JSONEncoder().encode(rows) else { return }
        try? data.write(to: url(), options: .atomic)
    }
}

extension TransferCenter {
    var persistSendQueue: Bool {
        settings.deviceName != "自检接收端" && !CommandLine.arguments.contains("--selftest")
    }

    func canQueue(_ peer: PeerDevice) -> Bool {
        peer.id != "phone-web" && !peer.id.isEmpty
    }

    func shouldQueue(_ peer: PeerDevice) -> Bool {
        var p = peer
        p.refreshOnline()
        return canQueue(p) && !p.online
    }

    func configureSender(deviceId: String, deviceName: String) {
        senderId = deviceId
        senderName = deviceName
    }

    func startRetryTimer() {
        retryTimer?.invalidate()
        retryTimer = Timer.scheduledTimer(withTimeInterval: 8, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.flushPending()
            }
        }
    }

    func restoreQueue() {
        guard persistSendQueue else { return }
        sendQueue = SendQueueFile.load()
        for job in sendQueue {
            if items.contains(where: { $0.id == job.itemId }) { continue }
            add(TransferItem(
                id: job.itemId, sessionId: job.sessionId, direction: .send, peerName: job.peerName,
                title: job.title, total: job.total, done: 0, speed: 0, state: .waitingPeer,
                detail: SendReachability.resumeHint, fileCount: job.fileCount,
                localPath: job.paths.first
            ))
        }
    }

    func enqueue(_ job: QueuedSend) {
        if let idx = sendQueue.firstIndex(where: { $0.itemId == job.itemId }) {
            sendQueue[idx] = job
        } else {
            sendQueue.append(job)
        }
        persistQueue()
    }

    func dropQueued(itemId: UUID) {
        sendQueue.removeAll { $0.itemId == itemId }
        persistQueue()
    }

    func persistQueue() {
        guard persistSendQueue else { return }
        SendQueueFile.save(sendQueue)
    }

    func flushPending() {
        let ids = Set(sendQueue.map(\.peerId))
        for id in ids {
            guard let peer = resolvedPeer(id: id, fallback: nil) else { continue }
            var live = peer
            live.refreshOnline()
            guard live.online else { continue }
            flush(to: live)
        }
    }

    func flush(to peer: PeerDevice) {
        guard peer.id != "phone-web" else { return }
        guard sendQueue.contains(where: { $0.peerId == peer.id }) else { return }
        guard !flushingPeers.contains(peer.id) else { return }
        flushingPeers.insert(peer.id)
        Task { @MainActor in
            defer { flushingPeers.remove(peer.id) }
            while true {
                guard let job = takeReadyJob(peerId: peer.id) else { break }
                if await flags.isCancelled(job.sessionId) { continue }
                var live = resolvedPeer(id: peer.id, fallback: peer) ?? peer
                live.refreshOnline()
                if !live.online {
                    enqueue(job)
                    break
                }
                await runQueued(job, peer: live)
            }
        }
    }

    private func resolvedPeer(id: String, fallback: PeerDevice?) -> PeerDevice? {
        if let found = livePeer?(id) { return found }
        return fallback
    }

    private func takeReadyJob(peerId: String) -> QueuedSend? {
        let now = Date()
        guard let idx = sendQueue.firstIndex(where: { job in
            job.peerId == peerId && now.timeIntervalSince(job.lastAttempt) >= 20
        }) else { return nil }
        let job = sendQueue.remove(at: idx)
        persistQueue()
        return job
    }

    func runQueued(_ job: QueuedSend, peer: PeerDevice) async {
        if !job.text.isEmpty {
            await deliverText(to: peer, text: job.text, job: job)
            return
        }
        let urls = job.paths.map { URL(fileURLWithPath: $0) }
        let missing = urls.contains { !FileManager.default.fileExists(atPath: $0.path) }
        if missing || urls.isEmpty {
            fail(peerName: job.peerName, message: "文件已经不在了", itemId: job.itemId)
            return
        }
        mutate(job.itemId) { item in
            var item = item
            item.state = .waiting
            item.detail = "正在连接 \(peer.host)"
            return item
        }
        let snap = Snapshot(settings)
        let deviceId = senderId
        let deviceName = senderName
        let cancelFlags = flags
        do {
            let scanned = try await Task.detached {
                try FileScanner.collect(urls: urls)
            }.value
            let files = scanned.enumerated().map {
                FileOffer(index: $0.offset, relativePath: $0.element.relative, size: $0.element.size, modified: $0.element.modified)
            }
            try await Task.detached {
                try await TransferPipes.send(
                    peer: peer, files: files, localURLs: scanned.map(\.url),
                    sessionId: job.sessionId, itemId: job.itemId,
                    deviceId: deviceId, deviceName: deviceName, settings: snap, flags: cancelFlags,
                    onUpdate: { itemId, mutate in
                        await TransferCenter.shared?.mutate(itemId, mutate)
                    }
                )
            }.value
        } catch {
            if await flags.isCancelled(job.sessionId) { return }
            if SendReachability.isUnreachable(error) && canQueue(peer) {
                var again = job
                again.lastAttempt = Date()
                again.host = peer.host
                again.port = peer.port
                enqueue(again)
                mutate(job.itemId) { item in
                    var item = item
                    item.state = .waitingPeer
                    item.detail = SendReachability.resumeHint
                    item.speed = 0
                    return item
                }
                return
            }
            fail(peerName: job.peerName, message: (error as? AppError)?.description ?? error.localizedDescription, itemId: job.itemId)
        }
    }

    func deliverText(to peer: PeerDevice, text: String, job: QueuedSend?) async {
        let port = settings.port
        let pin = settings.pin
        let deviceId = senderId
        let deviceName = senderName
        do {
            try await Task.detached {
                let conn = NetFactory.connect(host: peer.host, port: peer.port)
                try await conn.startAndWait()
                let io = try await WireIO.open(conn, asServer: false)
                try await io.sendJSON(.hello(ControlMessage.Hello(
                    deviceId: deviceId, name: deviceName, port: port,
                    httpPort: Ports.http(port), os: "macOS", appVersion: "1.0.0",
                    pin: pin.isEmpty ? nil : pin
                )))
                _ = try await io.receiveFrame(timeout: 15)
                try await io.sendJSON(.text(ControlMessage.TextPayload(sessionId: UUID().uuidString, body: text)))
                io.cancel()
            }.value
            if let job {
                mutate(job.itemId) { item in
                    var item = item
                    item.state = .completed
                    item.detail = "已发出"
                    item.speed = 0
                    return item
                }
            }
        } catch {
            if let job, await flags.isCancelled(job.sessionId) { return }
            if SendReachability.isUnreachable(error) && canQueue(peer) {
                if var again = job {
                    again.lastAttempt = Date()
                    again.host = peer.host
                    again.port = peer.port
                    enqueue(again)
                    mutate(again.itemId) { item in
                        var item = item
                        item.state = .waitingPeer
                        item.detail = SendReachability.resumeHint
                        return item
                    }
                } else {
                    parkText(to: peer, text: text)
                }
                return
            }
            fail(peerName: peer.name, message: "文字没发出去", itemId: job?.itemId)
        }
    }

    func parkFiles(to peer: PeerDevice, scanned: [(url: URL, relative: String, size: UInt64, modified: TimeInterval)], itemId: UUID, sessionId: String) {
        let files = scanned.enumerated().map {
            FileOffer(index: $0.offset, relativePath: $0.element.relative, size: $0.element.size, modified: $0.element.modified)
        }
        let total = files.reduce(UInt64(0)) { $0 + $1.size }
        let title = files.count == 1 ? files[0].relativePath : "\(files.count) 个文件"
        if !items.contains(where: { $0.id == itemId }) {
            add(TransferItem(
                id: itemId, sessionId: sessionId, direction: .send, peerName: peer.name,
                title: title, total: total, done: 0, speed: 0, state: .waitingPeer,
                detail: SendReachability.resumeHint, fileCount: files.count,
                localPath: scanned.first?.url.path
            ))
        } else {
            mutate(itemId) { item in
                var item = item
                item.state = .waitingPeer
                item.detail = SendReachability.resumeHint
                item.speed = 0
                return item
            }
        }
        enqueue(QueuedSend(
            itemId: itemId, sessionId: sessionId, peerId: peer.id, peerName: peer.name,
            host: peer.host, port: peer.port, paths: scanned.map(\.url.path), text: "",
            title: title, total: total, fileCount: files.count
        ))
    }

    func parkText(to peer: PeerDevice, text: String) {
        let itemId = UUID()
        let sessionId = UUID().uuidString
        let title = text.count > 36 ? String(text.prefix(36)) + "…" : text
        add(TransferItem(
            id: itemId, sessionId: sessionId, direction: .send, peerName: peer.name,
            title: title, total: 0, done: 0, speed: 0, state: .waitingPeer,
            detail: SendReachability.resumeHint, fileCount: 0
        ))
        enqueue(QueuedSend(
            itemId: itemId, sessionId: sessionId, peerId: peer.id, peerName: peer.name,
            host: peer.host, port: peer.port, paths: [], text: text,
            title: title, total: 0, fileCount: 0
        ))
    }
}
